### PR TITLE
AK: Simplify RefPtr again

### PR DESCRIPTION
**Pull request by tomuta**
Let's make `RefPtr` not thread-safe itself again (similar to the std library) as the overhead is just too much, considering most of our code is single-threaded and doesn't need these features.

- [ ] Fix weird linker error:
```
i686-pc-serenity/bin/ld: Userland/Libraries/LibGUI/CMakeFiles/LibGUI.dir/Button.cpp.o: relocation R_386_GOTOFF against undefined symbol `_ZN12_GLOBAL__N_119registration_WidgetE' can not be used when making a shared object
i686-pc-serenity/bin/ld: final link failed: bad value
```
- [ ] Figure out a solution for #6086

### CODE REVIEW

**Comment by bgianfo** [comment]
If the overhead of RefPtr is too much of a perf hit, isn't that a sign that we are missing important moves(..) / passing by RefPtr by value in too many places?

**Comment by tomuta** [comment]
> If the overhead of RefPtr is too much of a perf hit, isn't that a sign that we are missing important moves(..) / passing by RefPtr by value in too many places?

I don't think that's a problem, it's the loops with atomic locking of the pointer value inside that kills performance (see #4902)

File: Userland/Libraries/LibGUI/Widget.h [hunk 1]
```diff
@@ -39,9 +39,13 @@
 #include <LibGfx/Rect.h>
 #include <LibGfx/StandardCursor.h>
 
+namespace {
+    extern Core::ObjectClassRegistration registration_Widget;
```
Comment (tomuta) [inline Userland/Libraries/LibGUI/Widget.h:43]: Not sure why this is causing the link error:
```
i686-pc-serenity/bin/ld: Userland/Libraries/LibGUI/CMakeFiles/LibGUI.dir/Button.cpp.o: relocation R_386_GOTOFF against undefined symbol `_ZN12_GLOBAL__N_119registration_WidgetE' can not be used when making a shared object
i686-pc-serenity/bin/ld: final link failed: bad value
```

**Comment by bgianfo** [comment]
> > If the overhead of RefPtr is too much of a perf hit, isn't that a sign that we are missing important moves(..) / passing by RefPtr by value in too many places?
> 
> I don't think that's a problem, it's the loops with atomic locking of the pointer value inside that kills performance (see #4902)

I looked at this some more... referencing the issue you mentioned and I think move(...) of String is just completely messed up? Each `swap(..)` is just calling `String`'s move constructor / move operator which just forward the move to the internal `StringImpl` ... `StringImpl` doesn't implement a move constructor / move operator. `StringImpl` derives form `RefCounted<T>` which also doesn't implement move constructor / move operator. `RefCounted<T>` derives from `RefcountedBase`.. which is marked as `AK_MAKE_NONCOPYABLE` / `AK_MAKE_NONMOVABLE`. Is there really a reason why we can't implement a more optimized `move(..)` for strings? I don't see why it's not just detach from one object and attach to the other? The one already has a stable refcount.. why do we need to increment / decrement at all... we just need to move owership of the ref?

**Comment by bgianfo** [comment]
cc: @alimpfard in case they have ideas / I'm missing something.

**Comment by alimpfard** [comment]
`RefPtr<T>` seems to have a move ctor though? `RefPtr.h:165`
The actual cost here seems to be in `RefPtrTraits<T>::exchange`, which forces an strong CAS on everything, which is used in that move ctor

**Comment by alimpfard** [comment]
Honestly, I don't think we need to actually worry about the state of the moved-from RefPtr, how about a `is_moved_from` (or equivalent) field that we set on move instead of actually setting it to null?

**Comment by awesomekling** [comment]
> Honestly, I don't think we need to actually worry about the state of the moved-from RefPtr, how about a `is_moved_from` (or equivalent) field that we set on move instead of actually setting it to null?

Moving from `RefPtr` is a well-defined operation that nulls out the pointer. We do this in tons of places. Perhaps you are thinking of `NonnullRefPtr`?

**Comment by alimpfard** [comment]
What I'm saying is, I don't think we need to _actually_ set the bits to zero, we could instead set a flag and treat said RefPtr as null, at least, not in `leak_ref_raw()`

**Comment by awesomekling** [comment]
> What I'm saying is, I don't think we need to _actually_ set the bits to zero, we could instead set a flag and treat said RefPtr as null, at least, not in `leak_ref_raw()`

Where do you store that flag though? (without increasing the size of `RefPtr`)

**Comment by awesomekling** [comment]
Regardless, having implicit synchronization built into a smart pointer is weird and extremely heavy. As Tom indicates in the commit, we should simply use external synchronization like any other code would.

**Comment by alimpfard** [comment]
well if we're going to ignore the thread-safety, then that's fine, no need for safety guarantees when it's not meant to be safe lol
if not though, I don't think making it a bit larger is a big loss if we can gain some perf benefit from it by not synchronising the null state

**Pull request closed without merging**